source/opt/propagator.h

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_PROPAGATOR_H_
+#define LIBSPIRV_OPT_PROPAGATOR_H_
+
+#include <functional>
+#include <unordered_map>
+#include <unordered_set>
+#include <queue>
+
+#include "cfg.h"
+#include "ir_context.h"
+#include "module.h"
+
+namespace spvtools {
+namespace opt {
+
+// Represents a CFG control edge. TODO(dnovillo): Move it to ir::CFG and always
+// build it. Alternately, move it to IRContext and build it on-demand.
+struct Edge {
+  explicit Edge(const ir::BasicBlock* b1, const ir::BasicBlock* b2)
+      : src(b1), dst(b2) {}
+  const ir::BasicBlock* src;
+  const ir::BasicBlock* dst;
+  bool operator==(const Edge& o) const { return src == o.src && dst == o.dst; }
+};
+
+}  // namespace opt
+}  // namespace spvtools
+
+// Hashing operator for Edge. Needed to instantiate unordered sets of Edges.
+namespace std {
+template <>
+struct std::hash<spvtools::opt::Edge> {
```

Comment:
You are already in the namespace.
std::hash -> hash

Reply:
Done.